fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java

```diff
@@ -299,15 +301,212 @@ public Statistics visitLogicalFilter(LogicalFilter<? extends Plan> filter, Void
+                    .getStatistics(olapScan.getRelationId());
+            if (optStats.isPresent()) {
+                double actualRowCount = olapScan.getTable().getRowCountForNereids();
+                // if estimated mv rowCount is more than actual row count, fall back to base table stats
+                if (actualRowCount > optStats.get().getRowCount()) {
+                    return optStats.get();
+                }
+            }
+        }
+
+        StatisticsBuilder builder = new StatisticsBuilder();
+
+        // for system table or FeUt, use ColumnStatistic.UNKNOWN
+        if (StatisticConstants.isSystemTable(olapTable) || !FeConstants.enableInternalSchemaDb
+                || ConnectContext.get() == null
+                || ConnectContext.get().getSessionVariable().internalSession) {
+            for (Slot slot : olapScan.getOutput()) {
+                builder.putColumnStatistics(slot, ColumnStatistic.UNKNOWN);
+            }
+            setHasUnknownColStatsInStatementContext();
+            builder.setRowCount(olapTable.getRowCountForNereids());
+            return builder.build();
+        }
+
+        // for regression shape test, get row count from columnStats.count
+        if (ConnectContext.get() == null || !ConnectContext.get().getSessionVariable().enableStats) {
+            // get row count from any visible slotReference's colStats
+            double rowCount = 1;
+            for (Slot slot : olapScan.getOutput()) {
+                if (isVisibleSlotReference(slot)) {
+                    ColumnStatistic cache = getColumnStatistic(olapTable, slot.getName(),
+                            olapScan.getSelectedIndexIdForMV());
+                    rowCount = Math.max(rowCount, cache.count);
+                }
+                builder.putColumnStatistics(slot,
+                        new ColumnStatisticBuilder(ColumnStatistic.UNKNOWN).setCount(rowCount).build());
+            }
+            setHasUnknownColStatsInStatementContext();
+            return builder.setRowCount(rowCount).build();
+        }
+
+        // build Stats for olapScan
+        // if slot is not slotReference or is invisible, use UNKNOWN
+        List<SlotReference> outputSlotReferences = new ArrayList<>();
+        for (Slot slot : olapScan.getOutput()) {
+            if (isVisibleSlotReference(slot)) {
+                outputSlotReferences.add((SlotReference) slot);
+            } else {
+                builder.putColumnStatistics(slot, ColumnStatistic.UNKNOWN);
+            }
+        }
+        // build col stats for outputSlotReferences
+        if (!olapScan.getSelectedPartitionIds().isEmpty()) {
+            double rowCount = getSelectedPartitionRowCount(olapScan);
+            // if partition row count is not available, fallback to table stats
+            if (rowCount > 0) {
+                List<String> selectedPartitionNames = new ArrayList<>(olapScan.getSelectedPartitionIds().size());
```

Comment:
use Lists.newArrayListWithExpectedSize() to avoid resize. because list will resize when filling rate more than a threshold